File: cpp/cpp_offer/offerII115.cc
// C++
// leetcode offerII 115
// https://leetcode.cn/problems/ur2n8P/

class Solution {
public:
	int he[10009], e[10009], ne[10009], in[10009];
	int idx = 0;

	void add(int a, int b) {
		e[idx] = b;
		ne[idx] = he[a];
		he[a] = idx++;
		in[b]++;
	}

	void initMap() {
		memset(he, -1, sizeof(he));
		memset(in, 0, sizeof(in));
		idx = 0;
	}

    bool sequenceReconstruction(vector<int>& nums, vector<vector<int>>& sequences) {
    	initMap();
    	
    	int n = nums.size();

    	for(auto &se : sequences) {
    		for(int i = 1; i < se.size(); i++) {
    			add(se[i - 1], se[i]);
    		}
    	}

    	queue<int> q;
    	for (int i = 1; i <= n; i++) {
    		if (in[i] == 0)
    			q.push(i);
    	}

    	int loc = 0;
    	while(!q.empty()) {
    		if(q.size() != 1)
    			return false;

    		int t = q.front();
    		if (nums[loc++] != t)
    			return false;
    		q.pop();

    		for(int i = he[t]; i != -1; i = ne[i]) {
    			int j = e[i];
    			if (--in[j] == 0) {
    				q.push(j);
    			}
    		}
    	}
    	return true;
    }
};

File: Go/Go_offer/offerII115.go
// Go
// leetcode offerII 115
// https://leetcode.cn/problems/ur2n8P/

type Map struct {
	He, Ne, E, In []int
	Idx           int
}

func Constructor() Map {
	n := 10009
	he := make([]int, n)
	for i := 0; i < n; i++ {
		he[i] = -1
	}
	return Map{he, make([]int, n), make([]int, n), make([]int, n), 0}
}

func (m *Map) add(a, b int) {
	m.E[m.Idx] = b
	m.Ne[m.Idx] = m.He[a]
	m.He[a] = m.Idx
	m.Idx++
	m.In[b]++
}

func sequenceReconstruction(nums []int, sequences [][]int) bool {
	n := len(nums)
	m := Constructor()
	for i := 0; i < len(sequences); i++ {
		for j := 1; j < len(sequences[i]); j++ {
			m.add(sequences[i][j-1], sequences[i][j])
		}
	}

	q := make([]int, 0, 2)
	for i := 1; i <= n; i++ {
		if m.In[i] == 0 {
			q = append(q, i)
		}
	}

	loc := 0
	for len(q) != 0 {
		// fmt.Println("start loop size", len(q))
		if len(q) != 1 {
			return false
		}

		t := q[0]
		// fmt.Println(t)
		if nums[loc] != t {
			return false
		}
		loc++

		for i := m.He[t]; i != -1; i = m.Ne[i] {
			j := m.E[i]
			// fmt.Println("edge",j)
			m.In[j]--
			if m.In[j] == 0 {
				q = append(q, j)
			}
		}
		q = q[1:]
		// fmt.Println("end loop size", len(q))
	}
	return true
}